File: Two Pointers & Sliding Window/findLengthOfShortestSubarray1574.py
'''
  i  
                   -
    -     -   -
  -   - -   - 
- 
      l   r
      
concat left of left and right of right

1 2 3 10 4 2 3 5
      l    r
'''
class Solution:
    def findLengthOfShortestSubarray(self, arr: List[int]) -> int:
        n = len(arr)
        l, r = 0, n - 1
        
        while l < n - 1 and arr[l] <= arr[l+1]:
            l += 1
        if l == n - 1: 
            return 0
        while r >= l and arr[r-1] <= arr[r]:
            r -= 1
        if r == 0:
            return n - 1
        
        res = min(n - l - 1, r) # to remove
        i, j = 0, r
        while i <= l and j < n:
            # print(res, i, j, j - i)
            if arr[j] >= arr[i]:
                res = min(res, j - i - 1)
                i += 1
            else:
                j += 1
        return res
 
        # for loop, same
        for i in range(l + 1):
            if arr[i] <= arr[r]:
                res = min(res, r - i - 1)
            elif arr[i] > arr[r] and r < n - 1:
                r += 1
            else:
                print(r) # r = n - 1
                break
        return res
            
            

File: Two Pointers & Sliding Window/findLengthOfShortestSubarray1574.java
class Solution {
    public int findLengthOfShortestSubarray(int[] arr) {
        int n = arr.length;
        int s = 0, e = n -1;
        
        while (s < n - 1 && arr[s] <= arr[s+1]) s++;
        if (s == n - 1) return 0;
        
        while (e >= s && arr[e-1] <= arr[e]) e--;
        if (e == 0) return n - 1;
        
        int res = Math.min(n - 1 - s, e);
        
        int i = 0, j = e;
        while (i <= s && j < n) {
            if (arr[j] >= arr[i]) {
                System.out.println(res);
                res = Math.min(res, j - i - 1);
                i++;
            }else j++;
        }
        
        return res;
    }
}